cognitive/src/main/scala/com/microsoft/ml/spark/cognitive/DocumentTranslator.scala

```diff
@@ -0,0 +1,172 @@
+
+  val sourceLanguage = new ServiceParam[String](this, "sourceLanguage", "Language code." +
+    " If none is specified, we will perform auto detect on the document.")
+
+  def setSourceLanguage(v: String): this.type = setScalarParam(sourceLanguage, v)
+
+  def setSourceLanguageCol(v: String): this.type = setVectorParam(sourceLanguage, v)
+
+  val sourceUrl = new ServiceParam[String](this, "sourceUrl", "Location of the folder /" +
+    " container or single file with your documents.", isRequired = true)
+
+  def setSourceUrl(v: String): this.type = setScalarParam(sourceUrl, v)
+
+  def setSourceUrlCol(v: String): this.type = setVectorParam(sourceUrl, v)
+
+  val sourceStorageSource = new ServiceParam[String](this, "sourceStorageSource",
+    "Storage source of source input.")
+
+  def setSourceStorageSource(v: String): this.type = setScalarParam(sourceStorageSource, v)
+
+  def setSourceStorageSourceCol(v: String): this.type = setVectorParam(sourceStorageSource, v)
+
+  val storageType = new ServiceParam[String](this, "storageType", "Storage type of the input" +
+    " documents source string. Required for single document translation only.")
+
+  def setStorageType(v: String): this.type = setScalarParam(storageType, v)
+
+  def setStorageTypeCol(v: String): this.type = setVectorParam(storageType, v)
+
+  val targets = new ServiceParam[Seq[TargetInput]](this, "targets", "Destination for the" +
+    " finished translated documents.")
+
+  def setTargets(v: Seq[TargetInput]): this.type = setScalarParam(targets, v)
+
+  def setTargetsCol(v: String): this.type = setVectorParam(targets, v)
+
+  override protected def prepareEntity: Row => Option[AbstractHttpEntity] = {
+    def fetchGlossaries(row: Row): Option[Seq[Glossary]] = {
+      try {
+        Option(row.getSeq(1).asInstanceOf[Seq[Row]].map(
+          x => Glossary(x.getString(0), x.getString(1), Option(x.getString(2)), Option(x.getString(3)))
+        ))
+      } catch {
+        case _: NullPointerException => Option(row.getAs[Seq[Glossary]]("glossaries"))
+      }
+    }
+
+    r =>
+      Some(new StringEntity(
+        Map("inputs" -> Seq(
+          BatchRequest(source = SourceInput(
+            filter = Option(DocumentFilter(
+              prefix = getValueOpt(r, filterPrefix),
+              suffix = getValueOpt(r, filterSuffix))),
+            language = getValueOpt(r, sourceLanguage),
+            storageSource = getValueOpt(r, sourceStorageSource),
+            sourceUrl = getValue(r, sourceUrl)),
+            storageType = getValueOpt(r, storageType),
+            targets = getValue(r, targets).asInstanceOf[Seq[Row]].map(
+              row => TargetInput(Option(row.getString(0)),
+                fetchGlossaries(row),
+                row.getString(2), row.getString(3), Option(row.getString(4))))
+          ))).toJson.compactPrint, ContentType.APPLICATION_JSON))
+  }
+
+  private def queryForResult(key: Option[String],
```

Comment:
Are there any similarities that would allow us to abstract this and other async querying logic into the same function?

Reply:
Done.